File: app/src/main/java/org/openobservatory/ooniprobe/model/database/TestDescriptor.java
package org.openobservatory.ooniprobe.model.database;

import android.content.Context;
import android.graphics.Color;

import com.google.common.collect.Lists;
import com.raizlabs.android.dbflow.annotation.Column;
import com.raizlabs.android.dbflow.annotation.PrimaryKey;
import com.raizlabs.android.dbflow.annotation.Table;
import com.raizlabs.android.dbflow.sql.language.SQLite;
import com.raizlabs.android.dbflow.sql.language.Where;
import com.raizlabs.android.dbflow.structure.BaseModel;

import org.openobservatory.engine.OONIRunNettest;
import org.openobservatory.ooniprobe.common.AppDatabase;
import org.openobservatory.ooniprobe.common.LocaleUtils;
import org.openobservatory.ooniprobe.common.MapUtility;
import org.openobservatory.ooniprobe.common.ThirdPartyServices;
import org.openobservatory.ooniprobe.domain.MapConverter;
import org.openobservatory.ooniprobe.domain.NettestConverter;
import org.openobservatory.ooniprobe.test.suite.OONIRunSuite;
import org.openobservatory.ooniprobe.test.test.AbstractTest;
import org.openobservatory.ooniprobe.test.test.WebConnectivity;

import java.io.Serializable;
import java.util.Date;
import java.util.HashMap;
import java.util.List;

import javax.annotation.Nullable;

enum Animations {
    checkMark("anim/checkMark.json"),
    circumvention("anim/circumvention.json"),
    crossMark("anim/crossMark.json"),
    experimental("anim/experimental.json"),
    instant_messaging("anim/instant_messaging.json"),
    middle_boxes("anim/middle_boxes.json"),
    performance("anim/performance.json"),
    websites("anim/websites.json");

    private final String asset;

    Animations(String asset) {
        this.asset = asset;
    }

    @Override
    public String toString() {
        return this.asset;
    }
}

@Table(database = AppDatabase.class)
public class TestDescriptor extends BaseModel implements Serializable {
    @PrimaryKey()
    private long runId;

    @Column
    private String name;

    @Column(name = "name_intl", typeConverter = MapConverter.class)
    private HashMap nameIntl;

    @Column(name = "short_description")
    private String shortDescription;

    @Column(name = "short_description_intl", typeConverter = MapConverter.class)
    private HashMap shortDescriptionIntl;

    @Column
    private String description;

    @Column(name = "description_intl", typeConverter = MapConverter.class)
    private HashMap descriptionIntl;

    @Column
    private String icon;

    @Column
    private String color;

    @Column
    private String animation;

    @Column
    private String author;

    @Column
    private boolean archived;

    @Column(name = "auto_run")
    private boolean autoRun;

    @Column(name = "auto_update")
    private boolean autoUpdate;

    @Column(name = "creation_time")
    private Date creationTime;

    @Column(name = "translation_creation_time")
    private Date translationCreationTime;
    @Column(typeConverter = NettestConverter.class)
    private List nettests;

    public long getRunId() {
        return runId;
    }

    public void setRunId(long runId) {
        this.runId = runId;
    }

    public String getName() {
        return MapUtility.getOrDefaultCompat(nameIntl, LocaleUtils.sLocale.getLanguage(), name).toString();
    }

    public void setName(String name) {
        this.name = name;
    }

    public HashMap getNameIntl() {
        return nameIntl;
    }

    public void setNameIntl(HashMap nameIntl) {
        this.nameIntl = nameIntl;
    }

    public String getShortDescription() {
        return MapUtility.getOrDefaultCompat(shortDescriptionIntl, LocaleUtils.sLocale.getLanguage(), shortDescription).toString();
    }

    public void setShortDescription(String shortDescription) {
        this.shortDescription = shortDescription;
    }

    public HashMap getShortDescriptionIntl() {
        return shortDescriptionIntl;
    }

    public void setShortDescriptionIntl(HashMap shortDescriptionIntl) {
        this.shortDescriptionIntl = shortDescriptionIntl;
    }

    public String getDescription() {
        return MapUtility.getOrDefaultCompat(descriptionIntl, LocaleUtils.sLocale.getLanguage(), description).toString();
    }

    public void setDescription(String description) {
        this.description = description;
    }

    public HashMap getDescriptionIntl() {
        return descriptionIntl;
    }

    public void setDescriptionIntl(HashMap descriptionIntl) {
        this.descriptionIntl = descriptionIntl;
    }

    public String getIcon() {
        return icon;
    }

    public void setIcon(String icon) {
        this.icon = icon;
    }

    public String getColor() {
        return color != null ? color : "#495057";
    }

    public void setColor(String color) {
        this.color = color;
    }

    public String getAnimation() {
        return animation;
    }

    public void setAnimation(String animation) {
        this.animation = animation;
    }

    public int getParsedColor() {
        return Color.parseColor(getColor());
    }

    @Nullable
    public String getAnimationAsset() {
        try {
            return Animations.valueOf(animation).toString();
        }catch (Exception e){
            e.printStackTrace();
            ThirdPartyServices.logException(e);
            return null;
        }
    }

    public String getAuthor() {
        return author;
    }

    public void setAuthor(String author) {
        this.author = author;
    }

    public boolean isArchived() {
        return archived;
    }

    public void setArchived(boolean archived) {
        this.archived = archived;
    }

    public boolean isAutoRun() {
        return autoRun;
    }

    public void setAutoRun(boolean autoRun) {
        this.autoRun = autoRun;
    }

    public boolean isAutoUpdate() {
        return autoUpdate;
    }

    public void setAutoUpdate(boolean autoUpdate) {
        this.autoUpdate = autoUpdate;
    }

    public Date getCreationTime() {
        return creationTime;
    }

    public void setCreationTime(Date creationTime) {
        this.creationTime = creationTime;
    }

    public Date getTranslationCreationTime() {
        return translationCreationTime;
    }

    public void setTranslationCreationTime(Date translationCreationTime) {
        this.translationCreationTime = translationCreationTime;
    }

    public List getNettests() {
        return nettests;
    }

    public void setNettests(List nettests) {
        this.nettests = nettests;
    }

    public OONIRunSuite getTestSuite(Context context) {
        List<AbstractTest> tests = Lists.transform(
                (List<OONIRunNettest>)getNettests(),
                nettest -> {
                    AbstractTest test = AbstractTest.getTestByName(nettest.getName());
                    if (nettest.getName().equals(WebConnectivity.NAME)){
                        for (String url : nettest.getInputs())
                            Url.checkExistingUrl(url);
                    }
                    test.setInputs(nettest.getInputs());
                    return test;
                }
        );
        return new OONIRunSuite(
                context,
                this,
                tests.toArray(new AbstractTest[0])
        );
    }

    public static Where<TestDescriptor> selectAllAvailable() {
        return SQLite.select().from(TestDescriptor.class)
                .where(TestDescriptor_Table.archived.eq(false));
    }

    public boolean shouldUpdate(TestDescriptor updatedDescriptor) {
        return updatedDescriptor.creationTime.after(creationTime)
                || updatedDescriptor.translationCreationTime.after(translationCreationTime);
    }


	public static final class Builder {
        private long runId;
        private String name;
        private HashMap nameIntl;
        private String shortDescription;
        private HashMap shortDescriptionIntl;
        private String description;
        private HashMap descriptionIntl;
        private String icon;
        private String color;
        private String animation;
        private String author;
        private boolean archived;
        private boolean autoRun;
        private boolean autoUpdate;
        private Date creationTime;
        private Date translationCreationTime;
        private List nettests;

        private Builder() {
        }

        public static Builder aTestDescriptor() {
            return new Builder();
        }

        public Builder withRunId(long runId) {
            this.runId = runId;
            return this;
        }

        public Builder withName(String name) {
            this.name = name;
            return this;
        }

        public Builder withNameIntl(HashMap nameIntl) {
            this.nameIntl = nameIntl;
            return this;
        }

        public Builder withShortDescription(String shortDescription) {
            this.shortDescription = shortDescription;
            return this;
        }

        public Builder withShortDescriptionIntl(HashMap shortDescriptionIntl) {
            this.shortDescriptionIntl = shortDescriptionIntl;
            return this;
        }

        public Builder withDescription(String description) {
            this.description = description;
            return this;
        }

        public Builder withDescriptionIntl(HashMap descriptionIntl) {
            this.descriptionIntl = descriptionIntl;
            return this;
        }

        public Builder withIcon(String icon) {
            this.icon = icon;
            return this;
        }

        public Builder withColor(String color) {
            this.color = color;
            return this;
        }

        public Builder withAnimation(String animation) {
            this.animation = animation;
            return this;
        }

        public Builder withAuthor(String author) {
            this.author = author;
            return this;
        }

        public Builder withArchived(boolean archived) {
            this.archived = archived;
            return this;
        }

        public Builder withAutoRun(boolean autoRun) {
            this.autoRun = autoRun;
            return this;
        }

        public Builder withAutoUpdate(boolean autoUpdate) {
            this.autoUpdate = autoUpdate;
            return this;
        }

        public Builder withCreationTime(Date creationTime) {
            this.creationTime = creationTime;
            return this;
        }

        public Builder withTranslationCreationTime(Date translationCreationTime) {
            this.translationCreationTime = translationCreationTime;
            return this;
        }

        public Builder withNettests(List nettests) {
            this.nettests = nettests;
            return this;
        }

        public TestDescriptor build() {
            TestDescriptor testDescriptor = new TestDescriptor();
            testDescriptor.setRunId(runId);
            testDescriptor.setName(name);
            testDescriptor.setNameIntl(nameIntl);
            testDescriptor.setShortDescription(shortDescription);
            testDescriptor.setShortDescriptionIntl(shortDescriptionIntl);
            testDescriptor.setDescription(description);
            testDescriptor.setDescriptionIntl(descriptionIntl);
            testDescriptor.setIcon(icon);
            testDescriptor.setAuthor(author);
            testDescriptor.setArchived(archived);
            testDescriptor.setAutoRun(autoRun);
            testDescriptor.setAutoUpdate(autoUpdate);
            testDescriptor.setCreationTime(creationTime);
            testDescriptor.setTranslationCreationTime(translationCreationTime);
            testDescriptor.setNettests(nettests);
            return testDescriptor;
        }
    }
}

File: app/src/main/java/org/openobservatory/ooniprobe/model/database/TestDescriptor.kt
package org.openobservatory.ooniprobe.model.database

import android.content.Context
import android.graphics.Color
import com.google.gson.Gson
import com.google.gson.reflect.TypeToken
import com.raizlabs.android.dbflow.annotation.Column
import com.raizlabs.android.dbflow.annotation.PrimaryKey
import com.raizlabs.android.dbflow.annotation.Table
import com.raizlabs.android.dbflow.converter.TypeConverter
import com.raizlabs.android.dbflow.structure.BaseModel
import org.openobservatory.engine.BaseNettest
import org.openobservatory.engine.OONIRunNettest
import org.openobservatory.ooniprobe.R
import org.openobservatory.ooniprobe.activity.adddescriptor.adapter.GroupedItem
import org.openobservatory.ooniprobe.activity.runtests.models.ChildItem
import org.openobservatory.ooniprobe.activity.runtests.models.GroupItem
import org.openobservatory.ooniprobe.common.AbstractDescriptor
import org.openobservatory.ooniprobe.common.AppDatabase
import org.openobservatory.ooniprobe.common.LocaleUtils
import org.openobservatory.ooniprobe.common.PreferenceManager
import org.openobservatory.ooniprobe.common.resolveStatus
import java.io.Serializable
import java.util.Date
import com.raizlabs.android.dbflow.annotation.TypeConverter as TypeConverterAnnotation

@Table(database = AppDatabase::class)
class TestDescriptor(

    @PrimaryKey
    var runId: Long = 0,

    @Column
    var name: String = "",

    @Column(name = "short_description")
    var shortDescription: String = "",

    @Column
    var description: String = "",

    @Column
    var author: String = "",

    @Column(typeConverter = NettestConverter::class)
    var nettests: Any = emptyList<OONIRunNettest>(),

    @Column(name = "name_intl", typeConverter = MapConverter::class)
    var nameIntl: Any? = null,

    @Column(name = "short_description_intl", typeConverter = MapConverter::class)
    var shortDescriptionIntl: Any? = null,

    @Column(name = "description_intl", typeConverter = MapConverter::class)
    var descriptionIntl: Any? = null,

    @Column
    var icon: String? = null,

    @Column
    var color: String? = null,

    @Column
    var animation: String? = null,

    @Column(name = "expiration_date")
    var expirationDate: Date? = null,

    @Column(name = "date_created")
    var dateCreated: Date? = null,

    @Column(name = "date_updated")
    var dateUpdated: Date? = null,

    @Column
    var revision: String? = null,


    @Column(name = "previous_revision")
    var previousRevision: String? = null,

    @Column(name = "is_expired")
    var isExpired: Boolean? = false,

    @Column(name = "auto_run")
    var isAutoRun: Boolean = true,

    @Column(name = "auto_update")
    var isAutoUpdate: Boolean = false,

) : BaseModel(), Serializable {
    fun preferencePrefix(): String {
        return "${runId}_"
    }

    fun localizedName() :String {
        return nameIntl.getValueForKey(LocaleUtils.getLocale().language) ?: name
    }

    fun localizedShortDescription() :String {
        return shortDescriptionIntl.getValueForKey(LocaleUtils.getLocale().language) ?: shortDescription
    }

    fun localizedDescription() :String {
        return descriptionIntl.getValueForKey(LocaleUtils.getLocale().language) ?: description
    }
}

/**
 * Check if the test descriptor should be updated
 * @param updatedDescriptor The new descriptor
 * @return True if the descriptor should be updated, False otherwise
 */
fun TestDescriptor.shouldUpdate(updatedDescriptor: TestDescriptor): Boolean {
    return (updatedDescriptor.dateUpdated?.after(dateUpdated) ?: true)
}

private const val DESCRIPTOR_TEST_NAME = "ooni_run"

class InstalledDescriptor(
    var testDescriptor: TestDescriptor,
    var tags: List<String>? = null
) : AbstractDescriptor<BaseNettest>(
    name = DESCRIPTOR_TEST_NAME,
    title = testDescriptor.localizedName(),
    shortDescription = testDescriptor.localizedShortDescription(),
    description = testDescriptor.localizedDescription(),
    icon = testDescriptor.icon ?: "settings_icon",
    color = Color.parseColor(testDescriptor.color ?: "#495057"),
    animation = testDescriptor.animation,
    dataUsage = R.string.TestResults_NotAvailable,
    nettests = when (testDescriptor.nettests is List<*>) {
        true -> (testDescriptor.nettests as List<*>)
            .filterIsInstance<OONIRunNettest>()
            .map { nettest: OONIRunNettest ->
                return@map BaseNettest(
                    name = nettest.name,
                    inputs = nettest.inputs,
                )
            }

        false -> emptyList()
    },
    descriptor = testDescriptor) {

    override fun getRuntime(context: Context, preferenceManager: PreferenceManager): Int {
        return R.string.TestResults_NotAvailable
    }

    fun isUpdateAvailable(): Boolean {
        return tags?.contains("updated") ?: false
    }

    override fun toRunTestsGroupItem(preferenceManager: PreferenceManager): GroupItem {
        return GroupItem(
            selected = false,
            name = this.name,
            title = this.title,
            shortDescription = this.shortDescription,
            description = this.description,
            icon = this.icon,
            color = Color.parseColor(testDescriptor.color ?: "#495057"),
            animation = this.animation,
            dataUsage = this.dataUsage,
            nettests = this.nettests.map { nettest ->
                ChildItem(
                    selected = preferenceManager.resolveStatus(
                        name = nettest.name,
                        prefix = preferencePrefix(),
                    ), name = nettest.name, inputs = nettest.inputs
                )
            },
            descriptor = this.testDescriptor
        )
    }

}

fun TestDescriptor.getNettests(): List<OONIRunNettest> {
    return when (nettests is List<*>) {
        true -> (nettests as List<*>)
            .filterIsInstance<OONIRunNettest>()
            .map { nettest: OONIRunNettest ->
                return@map GroupedItem(
                    name = nettest.name,
                    inputs = nettest.inputs,
                    selected = true
                )
            }

        false -> emptyList()
    }
}

fun Any?.getValueForKey(language: String): String? {
    return if (this is Map<*, *>) {
        this[language] as String?
    } else {
        null
    }
}


@TypeConverterAnnotation
class MapConverter : TypeConverter<String, Any>() {
    override fun getDBValue(model: Any?): String? {
        return Gson().toJson(model)
    }

    override fun getModelValue(json: String): HashMap<String, String> {
        val gson = Gson()
        val type = object : TypeToken<HashMap<String, String>>() {}.type
        return gson.fromJson(json, type)
    }
}


@TypeConverterAnnotation
class NettestConverter : TypeConverter<String, Any>() {
    override fun getDBValue(model: Any): String = Gson().toJson(model)

    override fun getModelValue(data: String): List<*> = Gson().fromJson(
        data, Array<OONIRunNettest>::class.java
    ).toList()
}

class ITestDescriptor(

    var runId: Long = 0,

    var name: String = "",

    var shortDescription: String = "",

    var description: String = "",

    var author: String = "",


    var nettests: List<OONIRunNettest>? = emptyList(),

    var nameIntl: HashMap<String, String>? = null,

    var shortDescriptionIntl: HashMap<String, String>? = null,

    var descriptionIntl: HashMap<String, String>? = null,

    var icon: String? = null,

    var color: String? = null,

    var animation: String? = null,

    val expirationDate: Date? = null,

    val dateCreated: Date? = null,

    val dateUpdated: Date? = null,

    val revision: String? = null,

    var previousRevision: String? = null,

    val isExpired: Boolean? = false,

    var isAutoUpdate: Boolean = false,

) : Serializable {
    fun toTestDescriptor(): TestDescriptor {
        return TestDescriptor(
            runId = runId,
            name = name,
            shortDescription = shortDescription,
            description = description,
            author = author,
            nettests = nettests?: emptyList<OONIRunNettest>(),
            nameIntl = nameIntl,
            shortDescriptionIntl = shortDescriptionIntl,
            descriptionIntl = descriptionIntl,
            icon = icon,
            color = color,
            animation = animation,
            expirationDate = expirationDate,
            dateCreated = dateCreated,
            dateUpdated = dateUpdated,
            revision = revision,
            previousRevision = previousRevision,
            isExpired = isExpired,
            isAutoUpdate = isAutoUpdate
        )
    }
}
